test/integration/storageversion/gc_test.go

```diff
@@ -0,0 +1,224 @@
+		createTestStorageVersion(t, kubeclient, idA)
+		if err := wait.PollImmediate(100*time.Millisecond, 10*time.Second, func() (bool, error) {
+			_, err := kubeclient.InternalV1alpha1().StorageVersions().Get(
+				context.TODO(), svName, metav1.GetOptions{})
+			if apierrors.IsNotFound(err) {
+				return true, nil
+			}
+			return false, nil
+		}); err == nil {
+			t.Fatalf("unexpected valid storage version getting GC'ed")
+		}
+		sv, err := kubeclient.InternalV1alpha1().StorageVersions().Get(
+			context.TODO(), svName, metav1.GetOptions{})
+		if err != nil {
+			t.Fatalf("failed to retrieve valid storage version: %v", err)
+		}
+		if len(sv.Status.StorageVersions) != 1 {
+			t.Errorf("unexpected number of storage version entries, expected 1, got: %v",
+				sv.Status.StorageVersions)
+		}
+		expectedID := idA
+		if sv.Status.StorageVersions[0].APIServerID != expectedID {
+			t.Errorf("unexpected storage version entry id, expected %v, got: %v",
+				expectedID, sv.Status.StorageVersions[0].APIServerID)
+		}
+		if err := kubeclient.InternalV1alpha1().StorageVersions().Delete(
+			context.TODO(), svName, metav1.DeleteOptions{}); err != nil {
+			t.Fatalf("failed to cleanup valid storage version: %v", err)
+		}
+	})
+
+	t.Run("deleting an id should delete a storage version entry that it owns", func(t *testing.T) {
+		createTestStorageVersion(t, kubeclient, idA, idB)
+		assertStorageVersionEntries(t, kubeclient, 2, idA)
+		deleteTestAPIServerIdentityLease(t, kubeclient, idA)
+		assertStorageVersionEntries(t, kubeclient, 1, idB)
+	})
+
+	t.Run("deleting an id should delete a storage version object that it owns entirely", func(t *testing.T) {
+		deleteTestAPIServerIdentityLease(t, kubeclient, idB)
+		assertStorageVersionGC(t, kubeclient)
+	})
+}
+
+func createTestStorageVersion(t *testing.T, client kubernetes.Interface, ids ...string) {
+	sv := &apiserverinternalv1alpha1.StorageVersion{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: svName,
+		},
+		Status: apiserverinternalv1alpha1.StorageVersionStatus{
+			CommonEncodingVersion: pointer.StringPtr(encodingVersion),
+		},
+	}
+	for _, id := range ids {
+		v := apiserverinternalv1alpha1.ServerStorageVersion{
+			APIServerID:       id,
+			EncodingVersion:   encodingVersion,
+			DecodableVersions: []string{encodingVersion},
+		}
+		sv.Status.StorageVersions = append(sv.Status.StorageVersions, v)
+	}
+
+	createdSV, err := client.InternalV1alpha1().StorageVersions().Create(context.TODO(), sv, metav1.CreateOptions{})
+	if err != nil {
+		t.Fatalf("failed to create storage version %s: %v", svName, err)
+	}
+	// update the created sv with intended status
+	createdSV.Status = sv.Status
+	if _, err := client.InternalV1alpha1().StorageVersions().UpdateStatus(
+		context.TODO(), createdSV, metav1.UpdateOptions{}); err != nil {
+		t.Fatalf("failed to update store version status: %v", err)
+	}
+}
+
+func assertStorageVersionGC(t *testing.T, client kubernetes.Interface) {
```

Comment:
nit: maybe assertStorageVersionGone or assertStorageVersionDeleted is more self-explanatory?